File: aoc2023/day_2.py
from functools import reduce
import re

# part one
re_draw = re.compile(r"(?:(\d+) ((:?green|red|blue)))")
bag = dict(red=12, green=13, blue=14)

with open("./data/day_2.txt") as fin:
    check = lambda game: all(map(lambda match: int(match[1]) <= bag[match[2]], re_draw.finditer(game)))
    counter = sum(map(lambda ii_game: ii_game[0] * check(ii_game[1]), enumerate(fin, 1)))
print(counter)

# part two
power = 0
with open("./data/day_2.txt") as fin:
    for line in fin:
        game = map(lambda match: (int(match[1]), match[2]), re_draw.finditer(line))
        max_draw = dict(red=0, green=0.0, blue=0.0)
        for number, color in game:
            max_draw[color] = max(max_draw[color], number)
        power += reduce(lambda x, y: x * y, max_draw.values())

print(power)

File: rusty2023/src/bin/day2.rs
use regex::Regex;
use std::cmp;
use std::collections::HashMap;
use std::fs;
use std::io::{prelude::*, BufReader};

fn part_one() -> u32 {
    let file = fs::File::open("./data/day2.txt");
    let reader = BufReader::new(file.unwrap());

    let mut bag = HashMap::new();
    bag.insert("red", 12u32);
    bag.insert("green", 13u32);
    bag.insert("blue", 14u32);

    let re_sample = Regex::new(r"(\d+) ((?:red|blue|green))").unwrap();

    return reader
        .lines()
        .map(|line| line.expect("something bad happened"))
        .enumerate()
        .filter(|(_, game)| {
            re_sample
                .captures_iter(&game)
                .map(|c| c.extract())
                .all(|(_, [num_str, color])| num_str.parse::<u32>().unwrap() <= bag[color])
        })
        .map(|(ii, _)| u32::try_from(ii).unwrap() + 1)
        .sum();
}

fn part_two() -> u32 {
    let file = fs::File::open("./data/day2.txt");
    let reader = BufReader::new(file.unwrap());

    let re_sample = Regex::new(r"(\d+) ((?:red|blue|green))").unwrap();

    return reader
        .lines()
        .map(|line| line.expect("something bad happened"))
        .enumerate()
        .map(|(_, game)| {
            let mut bag = HashMap::new();
            bag.insert("red", 0u32);
            bag.insert("blue", 0u32);
            bag.insert("green", 0u32);
            re_sample
                .captures_iter(&game)
                .map(|c| c.extract())
                .for_each(|(_, [num_str, color])| {
                    bag.insert(
                        color,
                        cmp::max(
                            **bag.get(color).as_ref().unwrap(),
                            num_str.parse::<u32>().unwrap(),
                        ),
                    );
                });

            return bag.values().cloned().reduce(|x, y| x * y).unwrap();
        })
        .sum();
}

fn main() {
    println!("{}", part_one());
    println!("{}", part_two());
}
